Optimize the Linguistic String Search Operations

tarekgh opened the pull request:
This change target optimizing the string search operations (e.g. IndexOf, LastIndexOf, StartsWith, EndsWith) when using ICU. We used to create a new ICU search handle using `usearch_openFromCollator` everytime we call any of the search operations. The change here is to cache the search handles same way we cached Collation handles. The search handles are a little bit different as it needs to include the source string we need to search in and the pattern string we need to search for. This require the search handle should be used exclusively and not be shared between 2 operations in same time. Everytime we use the handle we'll set the source and the patterns strings to the handle to be able perform the operation correctly.

Caching the search handles improved the performance for IndexOf and LastIndexOf. When searching in long strings and the match pattern near the other edge of the search direction or no match found, then it shows perf improvement ranges `13% ~ 20%`. With shorter source strings the perf improvement is almost doubled. Also StartsWith/EndsWith perf is almost doubled regardless of the source string length. 


# Before #
|                      Method |       Mean |      Error |     StdDev |
|---------------------------- |-----------:|-----------:|-----------:|
|      TestMultipleStartsWith | 903.234 us | 17.9408 us | 40.4955 us |
|        TestMultipleEndsWith | 968.899 us | 19.2996 us | 38.0956 us |
|              MixLongIndexOf |  25.768 us |  0.5439 us |  1.5693 us |
|          MixLongLastIndexOf |  25.358 us |  0.5027 us |  1.3505 us |
|             MixShortIndexOf |   7.761 us |  0.1544 us |  0.3959 us |
|         MixShortLastIndexOf |   8.092 us |  0.1546 us |  0.3551 us |


# After #
|                      Method |       Mean |     Error |     StdDev | 
|---------------------------- |-----------:|----------:|-----------:|
|      TestMultipleStartsWith | 370.672 us | 7.1230 us | 17.4729 us |
|        TestMultipleEndsWith | 412.960 us | 7.9043 us | 19.2402 us |
|              MixLongIndexOf |  20.691 us | 0.4134 us |  1.1928 us |
|          MixLongLastIndexOf |  21.982 us | 0.6489 us |  1.8929 us |
|             MixShortIndexOf |   3.564 us | 0.0840 us |  0.2451 us |
|         MixShortLastIndexOf |   3.541 us | 0.0701 us |  0.1810 us |

tarekgh committed:
Optimize the Linguistic String Searcg Operations

tarekgh commented:
https://github.com/dotnet/runtime/issues/40942

CC @adamsitnik @DrewScoggins @stephentoub

tarekgh committed:
Fix small typo

DrewScoggins commented:
Should we bring these benchmarks into the performance suite? Seems like we should, at the very least to guard these wins for the future.

tarekgh commented:
@DrewScoggins 

> Should we bring these benchmarks into the performance suite? Seems like we should, at the very least to guard these wins for the future.

I believe these are already covered by the tests mentioned here https://github.com/dotnet/runtime/issues/40942#issuecomment-675056206. we can look at the new numbers coming from these tests after applying this change.

adamsitnik commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
+        }
+
+        UStringSearch* pNull = NULL;
+        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], *pSearchIterator, pNull))
+        {
+            usearch_close(*pSearchIterator);
+            *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+        }
```
should we avoid the extra step and just store `USED_STRING_SEARCH` in the cache here and exit immediately on success?

this would avoid the additional calls to `pal_atomic_cas_ptr`, `usearch_setText` and `usearch_setPattern` for the case when the `pSortHandle->searchIteratorPerOption[options]` is `NULL`

```suggestion
        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], USED_STRING_SEARCH, pNull))
        {
            usearch_close(*pSearchIterator);
            *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
        }
        else
        {
            return options;
        }
```

tarekgh replied:
This is possible but we'll have to set the text and the pattern before we return. I can refactor the code to do that.

tarekgh replied:
I'll try your suggestion.

adamsitnik commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
+    }
+    else
+    {
+        pSortHandle->searchIteratorPerOption[searchCacheSlot] = pSearchIterator;
```
I wonder what happens with the reference to `text` and `pattern` set via `usearch_setText` and `usearch_setPattern`in `GetSearchIteratorFromSortHandleUsingCollator`.

Are we risking a memory leak? Should the `text` and `pattern` be somehow set to null here?

tarekgh replied:
I don't think there is any risk here. GC has no knowledge about the native pointers is assigned in some cache. GC doesn't track any native pointer operation on the string. The string can be collected after the call return and the object is not reachable.

adamsitnik reviewed (commented):
The initial results look very promising. Could you please run also the [performance repo benchmarks](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/libraries/System.Globalization/StringSearch.cs) and share the results here?

My only concern is a possible memory leak for `text` and `pattern`:

```cs
string a = "aaaaaa";
string b = "bbbbbb";

new CultureInfo("pl-PL").CompareInfo.IsPrefix(a, b);

a = null;
b = null;

GC.Collect();

// is `a` or `b` somehow referenced by `UStringSearch` stored in `searchIteratorPerOption` ?
```

stephentoub commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
+        }
+
+        UStringSearch* pNull = NULL;
+        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], *pSearchIterator, pNull))
```
Did you consider instead just having each thread store a single cached handle in a [ThreadStatic]?

tarekgh replied:
Honestly no but I can see some issues doing that: 

- This will require a dictionary lookup with every call.
- If there are many created threads means more native allocation will occur and this may be a concern especially we don't free them.
- minor: Threadstatic still have some extra cost and we'll be required to pass extra parameter to the pinvoke calls. 

If you think it is worth to try, I can try prototype it but I am not sure the concerns I have with it is mitigated.

stephentoub commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
+    if (*pSearchIterator == USED_STRING_SEARCH || !pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], USED_STRING_SEARCH, *pSearchIterator))
+    {
+        // The handle is in use. Create a temporary new one.
+        *pSearchIterator = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, pErr);
```
We cache one sort handle per culture in managed code, right?  What does that mean for your measurements then, if only the first thread gets to use that cached sort handle's single cached search iterator?

tarekgh replied:
That is what I was saying the description above:

```
The search handles are a little bit different as it needs to include the source string we need to search in and the pattern string we need to search for. This require the search handle should be used exclusively and not be shared between 2 operations in same time.
```

The worst case when having a raise condition, we'll just create a new search handle as we already doing today.

danmosemsft replied:
Should we have tests to explicitly try to trigger that race condition? I guess test runs will inevitably test it by accident, but if there is a bug there it will be less obvious because it will look like random failures.

tarekgh replied:
I have manually tested the race condition by running the search operation a million times simultaneously on number of threads equals to the number of procs in my machine and everything worked fine. I hesitated to add the test to our test suites as my machine was not usable when running the test.

stephentoub replied:
> The worst case when having a raise condition, we'll just create a new search handle as we already doing today.

That's my point. It doesn't seem so much like a race condition to me. If a server is doing a lot of concurrent during processing, it could be that just one core gets these speedups and the rest don't, so the benchmarks aren't necessarily meaningful for a solution at scale. Have I misunderstood? That's why I was asking about an alternative that is more consistent.

tarekgh replied:
fair enough. I'll think about it more to look if I can avoid this situation. The first I can think about is instead of caching one instance, I can cache more instance when I cannot get a free one.

danmosemsft replied:
> by running the search operation a million times simultaneously

I guess I was thinking it could be pretty easy, if you start two threads that both do a large string search and are both released on an event. It will presumably hit the "not cached" case most of the time.

tarekgh replied:
I am currently changing the code to address @stephentoub feedback and will look adding a test as @danmosemsft suggesting too. Thanks for all feedback I am getting here.

danmosemsft commented:
Just curious, does this help Windows when using ICU?

tarekgh commented:
> Just curious, does this help Windows when using ICU?

Do you mean when directly using ICU on Windows and not from the managed code? if so, then no as the caller of the ICU is responsible for creating/opening the search handles. but it help if Windows has any code wrapping the search calls as we do in the managed.

tarekgh commented:
@adamsitnik here is the diff results for the search perf test. Interesting the showed slower cases ar emostely ordinal and shouldn't be affected by my changes. I'll try to look more on the test itself to try make it more reliable.


| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (en-US, Ord |      1.33 |            16.64 |            22.15 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (en- |      1.28 |            21.03 |            26.84 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.12 |           753.51 |           846.48 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (, No |      1.12 |           621.19 |           696.35 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (, IgnoreCa |      1.10 |           237.26 |           261.39 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (, N |      1.09 |            23.03 |            25.16 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (, I |      1.09 |          1576.73 |          1713.41 | several?|

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (en-US, I |      1.96 |         16315.89 |          8344.18 | several?|
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (en-US, N |      1.84 |         14848.83 |          8052.22 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, Ordi |      1.71 |            24.54 |            14.34 | several?|
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (en-US, I |      1.69 |         13450.49 |          7977.85 | several?|
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (en-US, Ign |      1.61 |         30757.98 |         19057.70 | bimodal |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (en- |      1.61 |         61902.51 |         38394.12 |         |
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (, None,  |      1.58 |         14617.98 |          9268.79 | several?|
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (pl-PL, N |      1.52 |         17192.49 |         11283.35 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.52 |         30131.81 |         19794.45 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, Igno |      1.47 |         22473.13 |         15252.63 |         |
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (, Ignore |      1.38 |         13604.81 |          9845.07 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.37 |         25252.33 |         18385.95 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (, No |      1.37 |         25538.87 |         18641.28 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.36 |         26316.24 |         19366.86 | bimodal |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (en-US, Non |      1.33 |         13164.94 |          9902.41 | bimodal |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (pl-P |      1.32 |         27810.19 |         21100.52 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (en |      1.31 |          1200.49 |           917.50 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (, Ig |      1.31 |         26029.60 |         19915.36 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (en- |      1.29 |          1867.69 |          1453.14 |         |
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (en-US, N |      1.25 |           732.24 |           585.58 | bimodal |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (en-US, Non |      1.25 |           254.69 |           204.00 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (en |      1.24 |         38068.26 |         30699.23 | several?|
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (, None, Tr |      1.21 |         11417.65 |          9412.09 | bimodal |
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (en |      1.21 |            25.88 |            21.40 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (en- |      1.20 |            25.66 |            21.39 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (pl-PL, Non |      1.18 |         12285.80 |         10439.04 | bimodal |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.17 |           724.08 |           616.77 | bimodal |
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (en |      1.17 |            13.57 |            11.59 | several?|
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (en- |      1.17 |          1738.64 |          1489.02 | bimodal |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, None |      1.16 |          6144.75 |          5288.90 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (,  |      1.15 |          1079.55 |           939.51 | bimodal |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (pl- |      1.15 |          1526.20 |          1332.11 | several?|
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (, N |      1.12 |          1767.83 |          1576.95 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.11 |           127.57 |           114.49 | bimodal |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (en-US, Ign |      1.11 |           247.76 |           222.57 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (,  |      1.11 |            24.26 |            21.85 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, None |      1.11 |           256.49 |           231.03 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (en |      1.11 |          1018.20 |           919.14 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (, None, Fal |      1.10 |           232.63 |           211.51 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, Igno |      1.10 |           224.82 |           204.43 | several?|
| System.Globalization.Tests.StringSearch.IndexOf_Word_NotFound(Options: (en-US, I |      1.10 |           802.85 |           732.26 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_SecondHalf(Options: (en-US, Ign |      1.09 |           221.54 |           203.14 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (, None, Tru |      1.08 |          5850.05 |          5400.00 |         |
| System.Globalization.Tests.StringSearch.LastIndexOf_Word_NotFound(Options: (en-U |      1.07 |           663.06 |           617.89 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, Ordi |      1.07 |            99.77 |            93.13 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (, IgnoreCas |      1.07 |           267.95 |           251.30 | several?|
| System.Globalization.Tests.StringSearch.IsPrefix_DifferentFirstChar(Options: (en |      1.06 |            22.72 |            21.37 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, Igno |      1.06 |          5558.41 |          5235.70 |         |
| System.Globalization.Tests.StringSearch.IsSuffix_DifferentLastChar(Options: (en- |      1.06 |            22.75 |            21.55 |         |
| System.Globalization.Tests.StringSearch.IsPrefix_FirstHalf(Options: (en-US, Igno |      1.05 |           252.56 |           240.76 |         |

danmosemsft commented:
> Do you mean when directly using ICU on Windows and not from the managed code? 

I meant when in "ICU mode" on Windows - we know that in some cases that was slower than in NLS mode.

tarekgh commented:
> I meant when in "ICU mode" on Windows - we know that in some cases that was slower than in NLS mode.

Yes, linguistic string compare and search operations are slower when using ICU than when using NLS. 

https://github.com/dotnet/runtime/issues/40942#issuecomment-675036985
https://github.com/dotnet/runtime/issues/40942#issuecomment-675056206

danmosemsft commented:
> Yes, linguistic string compare and search operations are slower when using ICU than when using NLS.

Right and would this approach help reduce that gap? I realize this PR touches Unix files. I'm probably misunderstanding how it fits together on Windows.

tarekgh commented:
> Right and would this approach help reduce that gap? I realize this PR touches Unix files. I'm probably misunderstanding how it fits together on Windows.

got it. the changes here is affecting Windows too. all the ICU code paths are used for Windows and Linux. sorry for not getting the question first time :-)

danmosemsft commented:
Ah, right. Maybe we want to move this library out of the `/Unix/` folder at some point..

tarekgh committed:
Apply Adam's suggestion

safern commented:
I was doing that in my PR when I made it build for Windows. But @jkotas suggested to instead leave it as is as we should probably remove the OS specific paths for native and drive what's included and what's not on the CMake files instead, just as coreclr does it.

jkotas commented:
Yep, I do not think it is useful to be encoding whether the native library builds on Windows, Unix or both in the directory path - for everything under `src\libraries\Native\`.

tarekgh committed:
Support caching multiple search handles

tarekgh commented:
@stephentoub I have addressed your feedback and now I can allow caching more search handles and that should scale well. 

@danmosemsft I have added a test for exercising the code across multiple threads to test the cache accessibility in raise conditions.

@adamsitnik @DrewScoggins I have been using the StringSearch tests and I am not noticing it is giving m different results when running it multiple times on same machine and with the exact same runtime bits (the base and the updated). I have locally changed the test a little to add a loop in every case to get more reasonable results and I think this helped and showing better results. I am pasting the results here but most the slower cases are ordinal which shouldn't be affected by my changes at all but the test showed it slower for some reason. Note, I changed the namespace, class name, and the method names just be be short and fit in the diff result.

| Slower                                                       | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| ------------------------------------------------------------ | ---------:| ----------------:| ----------------:| -------- |
| SGT.SS.IndexOf_Word_NF(Options: (en-US, Ordinal, False))     |      1.43 |         45937.12 |         65885.34 | bimodal |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, Ordinal, False)) |      1.18 |        115238.33 |        135767.05 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, None, True))    |      1.05 |        861251.74 |        905499.82 |         |

| Faster                                                                  | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ----------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| SGT.SS.IndexOf_Word_NF(Options: (, None, True))                         |      2.03 |      14149660.94 |       6985486.46 |         |
| SGT.SS.IndexOf_Word_NF(Options: (pl-PL, None, False))                   |      1.94 |      18258180.77 |       9410175.00 | bimodal |
| SGT.SS.IndexOf_Word_NF(Options: (en-US, IgnoreCase, True))              |      1.77 |      13161115.63 |       7450892.71 |         |
| SGT.SS.IndexOf_Word_NF(Options: (en-US, IgnoreSymbols, False))          |      1.76 |      12619739.06 |       7174940.63 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (pl-PL, None, False))               |      1.75 |      29810390.00 |      16993313.33 | several?|
| SGT.SS.IndexOf_Word_NF(Options: (en-US, None, True))                    |      1.71 |      12538931.25 |       7343355.21 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, IgnoreCase, True))          |      1.67 |      26948283.33 |      16151859.09 |         |
| SGT.SS.IndexOf_Word_NF(Options: (, IgnoreCase, True))                   |      1.55 |      11212507.81 |       7213828.13 |         |
| SGT.SS.IsPrefix_FH(Options: (en-US, IgnoreSymbols, False))              |      1.54 |      21852103.85 |      14145925.00 |         |
| SGT.SS.IsSuffix_DifferentLC(Options: (en-US, Ordinal, False))           |      1.47 |         27513.06 |         18709.30 |         |
| SGT.SS.IsSuffix_SH(Options: (en-US, IgnoreSymbols, False))              |      1.46 |      24717088.89 |      16901530.00 | several?|
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, None, True))                |      1.46 |      22570716.67 |      15499031.25 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, IgnoreSymbols, False))      |      1.44 |      21487418.18 |      14898540.63 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (, None, True))                     |      1.39 |      23055127.27 |      16530726.67 |         |
| SGT.SS.IsSuffix_SH(Options: (, None, True))                             |      1.39 |      10734775.00 |       7732342.19 | bimodal |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, IgnoreSymbols, False))     |      1.37 |      36537300.00 |      26735330.00 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (, IgnoreCase, True))               |      1.36 |      21142226.92 |      15575481.25 |         |
| SGT.SS.IsPrefix_FH(Options: (pl-PL, None, False))                       |      1.27 |       7814503.13 |       6172872.92 |         |
| SGT.SS.IsSuffix_DifferentLC(Options: (en-US, OrdinalIgnoreCase, False)) |      1.26 |        189689.91 |        150495.77 | several?|
| SGT.SS.IsSuffix_DifferentLC(Options: (pl-PL, None, False))              |      1.26 |       1541743.75 |       1226774.11 |         |
| SGT.SS.IsSuffix_SH(Options: (en-US, IgnoreNonSpace, False))             |      1.25 |        235799.80 |        188080.40 |         |
| SGT.SS.IsSuffix_DifferentLC(Options: (en-US, IgnoreSymbols, False))     |      1.22 |      42056583.33 |      34420942.86 |         |
| SGT.SS.IndexOf_Word_NF(Options: (en-US, OrdinalIgnoreCase, False))      |      1.20 |        780215.63 |        652637.64 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, Ordinal, False))           |      1.18 |         10207.88 |          8632.97 |         |
| SGT.SS.IsPrefix_FH(Options: (en-US, Ordinal, False))                    |      1.18 |         14315.65 |         12122.36 |         |
| SGT.SS.IsSuffix_SH(Options: (en-US, None, True))                        |      1.18 |       9688581.25 |       8233521.88 | bimodal |
| SGT.SS.IsPrefix_DifferentFC(Options: (, None, False))                   |      1.17 |         22354.71 |         19151.02 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, IgnoreCase, True))         |      1.15 |        984311.67 |        854599.22 |         |
| SGT.SS.IsPrefix_FH(Options: (en-US, IgnoreNonSpace, False))             |      1.14 |        210044.43 |        183514.14 |         |
| SGT.SS.IsPrefix_FH(Options: (, None, True))                             |      1.14 |       5359906.25 |       4695378.91 |         |
| SGT.SS.IndexOf_Word_NF(Options: (en-US, IgnoreNonSpace, False))         |      1.12 |        601908.85 |        536391.04 | several?|
| SGT.SS.IsSuffix_DifferentLC(Options: (en-US, IgnoreCase, False))        |      1.12 |         24030.77 |         21459.00 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (, None, True))                    |      1.11 |        902719.57 |        811764.93 |         |
| SGT.SS.IsSuffix_DifferentLC(Options: (en-US, None, False))              |      1.11 |         21778.02 |         19601.82 | bimodal |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, IgnoreCase, False))         |      1.10 |        732113.80 |        664367.75 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, OrdinalIgnoreCase, False))  |      1.10 |        780478.32 |        709557.24 |         |
| SGT.SS.IsSuffix_SH(Options: (en-US, OrdinalIgnoreCase, False))          |      1.09 |         98611.76 |         90069.54 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, OrdinalIgnoreCase, False)) |      1.08 |         11739.42 |         10824.99 | several?|
| SGT.SS.IsPrefix_FH(Options: (en-US, None, True))                        |      1.08 |       5108093.75 |       4717670.31 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (pl-PL, None, False))              |      1.08 |        879192.60 |        812896.71 |         |
| SGT.SS.IndexOf_Word_NF(Options: (, None, False))                        |      1.08 |        572535.21 |        530466.98 | bimodal |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, IgnoreNonSpace, False))     |      1.08 |        584718.85 |        541787.28 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, None, False))              |      1.07 |         20389.91 |         19001.52 |         |
| SGT.SS.IsSuffix_SH(Options: (pl-PL, None, False))                       |      1.07 |       9538002.27 |       8912203.13 |         |
| SGT.SS.IndexOf_Word_NF(Options: (en-US, None, False))                   |      1.06 |        570469.58 |        536467.13 |         |
| SGT.SS.IsSuffix_SH(Options: (en-US, None, False))                       |      1.06 |        197056.06 |        185795.68 |         |
| SGT.SS.IsPrefix_DifferentFC(Options: (en-US, IgnoreNonSpace, False))    |      1.06 |         19985.05 |         18921.23 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (, None, False))                    |      1.05 |        570901.67 |        544106.35 |         |
| SGT.SS.LastIndexOf_Word_NF(Options: (en-US, None, False))               |      1.05 |        573550.12 |        548621.67 |         |
| SGT.SS.IsPrefix_FH(Options: (en-US, OrdinalIgnoreCase, False))          |      1.03 |         94122.27 |         91274.20 |         |

tarekgh committed:
Fix Linux  build break

am11 commented on src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:
```diff
@@ -38,8 +38,16 @@ c_static_assert_msg(USEARCH_DONE == -1, "managed side requires -1 for not found"
+
 typedef struct { int32_t key; UCollator* UCollator; } TCollatorMap;
 
+typedef struct SEARCHNODE
+{
+    UStringSearch* searchIterator;
+    struct SEARCHNODE* next;
+} SearchIteratorNode;
```
a nit: it can be simplified to:

```c
typedef struct SearchIteratorNode
{
    int* searchIterator;
    struct SearchIteratorNode* next;
};
```

tarekgh replied:
@am11, your suggestion causing build breaks like the following:

```
 /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:45:1: error: typedef requires a name [-Werror,-Wmissing-declarations]
  typedef struct SearchIteratorNode
  ^~~~~~~
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:59:5: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
      SearchIteratorNode searchIteratorList[CompareOptionsMask + 1];
      ^
      struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:396:17: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
                  SearchIteratorNode* pNext = pSortHandle->searchIteratorList[i].next;
                  ^
                  struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:405:21: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
                      SearchIteratorNode* pCurrent = pNext;
                      ^
                      struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:451:5: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
      SearchIteratorNode* node = (SearchIteratorNode*) malloc(sizeof(SearchIteratorNode));
      ^
      struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:451:52: error: expected expression
      SearchIteratorNode* node = (SearchIteratorNode*) malloc(sizeof(SearchIteratorNode));
                                                     ^
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:451:33: error: use of undeclared identifier 'SearchIteratorNode'
      SearchIteratorNode* node = (SearchIteratorNode*) malloc(sizeof(SearchIteratorNode));
                                  ^
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:460:5: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
      SearchIteratorNode* pCurrent = &pSortHandle->searchIteratorList[options];
      ^
      struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:463:5: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
      SearchIteratorNode* pNull = NULL;
      ^
      struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:483:5: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
      SearchIteratorNode* pCurrent = &pSortHandle->searchIteratorList[options];
      ^
      struct
  /root/oss/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:536:5: error: must use 'struct' tag to refer to type 'SearchIteratorNode'
      SearchIteratorNode* pCurrent = &pSortHandle->searchIteratorList[options];
      ^
      struct
```

am11 replied:
Ops, fixed the typo in alias.

tarekgh committed:
address feedback

tarekgh commented:
@danmosemsft @stephentoub @adamsitnik do you have any more feedback?

danmosemsft commented:
Not me

tarekgh commented:
@safern could you please have a look when you have time?

safern reviewed (approved):
LGTM. The test that is added is marked as `Outerloop`... should we run the outerloop tests in this PR to make sure it is green in all flavors?

tarekgh commented:
/azp run runtime-libraries outerloop

tarekgh commented:
@safern do you know what https://github.com/dotnet/runtime/pull/43065#issuecomment-708087497 means? what should I do?

safern commented:
Ah is because the pipeline name now has `coreclr` in it... let me trigger it.

safern commented:
/azp run runtime-libraries-coreclr outerloop

tarekgh commented:
All outer loop test failures are unrelated and my added outer loop test not failed.
